### PR TITLE
core: add state transition for blob transaction

### PR DESCRIPTION
**Pull request by andicrypt**
This pr adds BLOBHASH, BLOBBASEFEE opcodes, kzg point evaluation precompile and handles state transition of blob transaction including: adding blob fields to transaction message and receipt, verify state transition for block and in pre-check, transfer blob fee to Ronin treasury address.

### CODE REVIEW

File: core/state_processor.go [hunk 1]
```diff
@@ -234,6 +234,6 @@ func ApplyTransaction(
 	}
 	// Create a new context to be used in the EVM environment
 	blockContext := NewEVMBlockContext(header, bc, author, publishEvents...)
-	vmenv := vm.NewEVM(blockContext, vm.TxContext{}, statedb, config, cfg)
+	vmenv := vm.NewEVM(blockContext, vm.TxContext{BlobHashes: tx.BlobHashes()}, statedb, config, cfg)
```
Comment (minh-bq) [inline core/state_processor.go:240]: This seems redundant, in `applyTransaction` the new tx context is created this is overwritten by the newly created one.

File: core/state_transition.go [hunk 2]
```diff
@@ -226,25 +227,42 @@ func (st *StateTransition) buyGas() error {
...
+				blobBalanceCheck.Mul(blobBalanceCheck, st.msg.BlobGasFeeCap())
+				balanceCheck.Add(balanceCheck, blobBalanceCheck)
+
+				// Pay for blobGasUsed * actual blob fee
+				blobFee = new(big.Int).SetUint64(blobGas)
+				blobFee.Mul(blobFee, st.evm.Context.BlobBaseFee)
+				effectiveGasFee.Add(effectiveGasFee, blobFee)
+			}
+		}
+		balanceCheck := new(big.Int).Add(balanceCheck, st.value)
```
Comment (minh-bq) [inline core/state_transition.go:268]: balanceCheck.Add(balanceCheck, st.value)

File: core/state_transition.go [hunk 3]
```diff
@@ -217,67 +221,95 @@ func (st *StateTransition) to() common.Address {
...
 	}
-	st.gas += st.msg.Gas()
+	st.gas += msg.Gas()
+
+	st.initialGas = msg.Gas()
 
-	st.initialGas = st.msg.Gas()
+	// Transfer blob gas fee to Ronin treasury address. If the blob tx fails,
+	// the fee will not be refund.
+	if st.blobGasUsed() > 0 {
```
Comment (minh-bq) [inline core/state_transition.go:283]: Should check if blobFee > 0

File: core/vm/contracts.go [hunk 4]
```diff
@@ -305,9 +320,10 @@ var (
 // modexpMultComplexity implements bigModexp multComplexity formula, as defined in EIP-198
 //
 // def mult_complexity(x):
-//    if x <= 64: return x ** 2
-//    elif x <= 1024: return x ** 2 // 4 + 96 * x - 3072
-//    else: return x ** 2 // 16 + 480 * x - 199680
+//
```
Comment (minh-bq) [inline core/vm/contracts.go:323]: Don't commit this chunk

**Review by minh-bq — COMMENTED** [review]
Should add more testcases for balance check after state transition

**Comment by andicrypt** [comment]
> Should add more testcases for balance check after state transition

Added!

**Review by minh-bq — APPROVED** [review]

**Review by Francesco4203 — APPROVED** [review]

**Pull request merged**